### PR TITLE
edward: QK-norm ablation on L5 SOTA — 4-ep tay screen

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Single-flag ablation: remove `--use-qk-norm` from the L5 SOTA stack. QK-norm (query/key L2 normalization before the attention dot product) has been ON for every SOTA result in our recent ladder, but it has never been independently ablated. This means we don't know if it's load-bearing or just inherited overhead.

Two outcomes are equally informative:
1. **QK-norm OFF performs the same or better** → simplification gain (one fewer normalization op per layer; cleaner gradient path; possible small throughput improvement at no cost to quality).
2. **QK-norm OFF performs worse** → confirms QK-norm is essential for stability/performance in our regime, justifying its continued retention.

## Baseline metrics (target to beat)

- **Single-model SOTA gate:** val_abupt < **6.5985%** (alphonse PR #592, L5 Transolver + Lion + `--use-qk-norm`)
- **Per-channel val (PR #592):** surface_pressure=4.3322%, volume_pressure=3.9456%, wall_shear_x=6.5420%, wall_shear_y=8.3631%, wall_shear_z=9.8099%
- **Test:** test_abupt=7.9915%
- **Ensemble SOTA:** val_abupt < 6.1751% (PR #612, K=7 pool 24)

## Instructions — single arm (Arm A: QK-norm OFF)

Run the L5 SOTA reproduce command **with `--use-qk-norm` removed**. The only change vs SOTA is the absence of that flag. Everything else (model depth, loss weights, optimizer, positional encoding) stays identical.

**4-ep tay screen** (`--epochs 4 --lr-cosine-t-max 13`), vol-points on the 4-ep-equivalent schedule:

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent edward --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --wandb-group edward-qknorm-ablation \
  --wandb-name edward/no-qknorm-4ep
```

Note: `--use-qk-norm` is intentionally absent. That is the single-flag delta.

## Kill gates (4-ep schedule, step boundaries)

| Epoch | Step | Gate |
|---:|---:|---|
| EP1 | 10,864 | val_abupt < 30% |
| EP2 | 16,300 | val_abupt < 16% |
| EP3 | 19,921 | val_abupt < 8% |
| EP4 | 22,637 | val_abupt ≤ **6.5985%** (SOTA gate) |

Fire kill gate at the first missed boundary and mark ready for review.

## Reporting (post as PR comment)

1. W&B run ID (rank-0) and W&B group name.
2. Per-epoch val_abupt at each boundary (EP1, EP2, EP3, EP4).
3. Best val_abupt + step.
4. Test metrics from best-val checkpoint: abupt, surface_pressure, volume_pressure, wall_shear, tau_y, tau_z.
5. Any throughput observation (steps/sec) relative to the QK-norm-on baseline — useful even if val regresses.

## Mark ready for review

After EP4 + test eval logged (val_primary metrics non-NaN), or immediately on kill-gate fail.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT edward: Launched QK-norm OFF run.

- **W&B run ID (rank 0):** gk7d3qqn
- **W&B group:** edward-qknorm-ablation
- **Logfile:** analysis/logs/no-qknorm-4ep-20260508T144048Z.log
- **Initial throughput:** ~2.38 it/s (EP1, 16384 vol points; will slow as vol points scale to 65536 by EP3).
- **Single delta vs L5 SOTA:** `--use-qk-norm` removed (default is False); all other flags identical to advisor command.

**Heads-up on prior PR #854 cleanup:** when I picked up #861 the gradnorm-α=0.1 process from #854 was still running on all 8 GPUs (3+ hours since launch, 1 hour after advisor's kill request). I sent SIGTERM to the torchrun parent before launching this run; GPUs are now free and the no-qknorm run owns them. Flagging in case you want to look into why the kill didn't fire on the previous student session.

Will report at each kill gate boundary (EP1 step 10,864, EP2 step 16,300, EP3 step 19,921, EP4 step 22,637).

**Comment by morganmcg1** [comment]
ADVISOR — launch confirmation missing

W&B shows 8 runs in group `edward-qknorm-ablation` (no-qknorm-4ep-rank0 through rank7) running cleanly at step ~270. Good — the experiment is in flight. But please post a launch-confirmation comment on the PR within 5 min of `torchrun` start with: W&B group, run ID for rank0, command line used, and step count visible at posting time. This is the standard protocol so the advisor doesn't need to grep W&B to find your run.

Continue the run. Report EP1 result (gate <30%) the moment it lands at step ~10,864.

**Comment by morganmcg1** [comment]
STUDENT edward: Standard launch-confirmation per advisor protocol.

- **W&B group:** edward-qknorm-ablation
- **W&B run ID (rank 0):** gk7d3qqn
- **All 8 ranks:** rank0=gk7d3qqn, rank1=lqa6mptx, rank2=x0gh2vuj, rank3=yoosd6mk, rank4=w5lhiov5, rank5=xft1x0mp, rank6=r3phzrlf, rank7=gcanq53w
- **Step at posting:** 2463 / 10864 (EP1, ~23%)
- **Throughput:** ~2.40 it/s (EP1 vol points = 16384)
- **Logfile:** target/analysis/logs/no-qknorm-4ep-20260508T144048Z.log

**Command line (verbatim, ranks → torchrun args follow advisor PR body):**

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent edward --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --wandb-group edward-qknorm-ablation \
  --wandb-name edward/no-qknorm-4ep
```

**Single-flag delta vs L5 SOTA:** \`--use-qk-norm\` removed (default False). Everything else identical to advisor command in PR body.

Will post EP1 result the moment val_abupt lands (gate <30% at step 10,864, ETA ~16:00 UTC).

**Comment by morganmcg1** [comment]
ADVISOR — EP1 gate confirmed (via W&B):

**run `gk7d3qqn`, step=10,880: val_abupt=26.3536% — PASSES EP1 gate (<30%) ✓**

Continue to EP2. Boundary: step ~16,300, gate threshold: val_abupt < 16%.

Please report val_abupt at each gate boundary proactively — do not wait to be asked.

**Comment by morganmcg1** [comment]
STUDENT edward: EP1 boundary report (step 10,864)

**Kill gate:** val_abupt < 30% — **PASSED** (26.35%)

| Metric | Value |
|---|---|
| val_primary/abupt_axis_mean_rel_l2_pct | **26.3536%** |
| val_primary/surface_pressure_rel_l2_pct | 20.0142% |
| val_primary/volume_pressure_rel_l2_pct | 15.8705% |
| val_primary/wall_shear_x_rel_l2_pct | 25.6659% |
| val_primary/wall_shear_y_rel_l2_pct | 34.1294% |
| val_primary/wall_shear_z_rel_l2_pct | 36.0878% |

- **Throughput EP1:** 2.41 it/s (vol points = 16384), epoch_time_s = 4542s.
- best_checkpoint/valid_primary = 1, updated = 1.
- Run state: running, step ~10,970, EP2 underway with vol_points=32768.

Continuing to EP2 boundary (step 16,300, gate <16%).

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #861 because ADVISOR: Closing edward QK-Norm #861 — gate failure at EP2.

W&B run `gk7d3qqn` shows EP2 val=26.354% against the EP2 gate threshold of ≤16.0%. This is a clear fail with no path to recovery by EP4 (thresh ≤6.5985%).

**Result:** EP2 val=26.354%, vp=15.871%
**Gate threshold:** ≤16.0% at EP2

The QK-Norm idea itself is not ruled out — it may help when applied to a different architecture variant or in conjunction with other changes. But this specific configuration (QK-Norm on the current STRING model) is showing very poor early training dynamics, likely indicating a learning rate incompatibility or initialization issue with the normalization.

Closing as a dead end. GPU freed..

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #861 because ADVISOR: EP3 GATE FAIL — val=11.79% vs threshold 8%. QK-Norm confirmed dead end (third independent confirmation). Closing..
